sql/analyzer/memo.go

```diff
@@ -363,6 +381,92 @@ func (p *relProps) InputTables() sql.FastIntSet {
 	return p.inputTables
 }
 
+// sortedOnDistinct returns true if a relation's inputs are sorted on the
+// full output schema. The OrderedDistinct operator can be used in this
+// case.
+func sortedOnDistinct(rel relExpr) bool {
+	switch r := rel.(type) {
+	case *max1Row:
+		return true
+	case *project:
+		if _, ok := r.child.best.(*max1Row); ok {
+			return true
+		}
+		sortedOn := sortedColsForRel(r.child.best)
+		childOutputs := r.outputCols()
+		if len(sortedOn) < len(childOutputs) {
+			return false
+		}
+		sorted := make(map[tableCol]struct{})
+		for _, c := range sortedOn {
+			sorted[tableCol{table: c.Source, col: c.Name}] = struct{}{}
+		}
+		for _, c := range childOutputs {
+			if _, ok := sorted[tableCol{table: strings.ToLower(c.Source), col: strings.ToLower(c.Name)}]; !ok {
+				return false
+			}
+		}
+		return true
+	default:
+		return false
+	}
+}
+
+func sortedColsForRel(rel relExpr) sql.Schema {
+	switch r := rel.(type) {
+	case *tableScan:
+		tab, ok := r.table.Table.(sql.PrimaryKeyTable)
+		if ok {
+			ords := tab.PrimaryKeySchema().PkOrdinals
+			var pks sql.Schema
+			for _, i := range ords {
+				pks = append(pks, tab.PrimaryKeySchema().Schema[i])
+			}
+			return pks
+		}
+	case *mergeJoin:
+		var ret sql.Schema
+		for _, e := range r.innerScan.idx.Expressions() {
+			parts := strings.Split(e, ".")
```

Comment:
This is fraught, column names can in fact include a "."

Not much you can do right now, but note it